### PR TITLE
    Stability fixes and optimizations

### PR DESCRIPTION
**Pull request by santanusinha**
- Introduced communication execption
- Introduced communicator for http
- Introduced circuit breakers for drove and http
- Staleness check is done centrally at ServiceRegistryUpdater
- Bugfix: If upstream is inactive, the refresher will override timestamps with current time so that all nodes don't go stale in one minute
- Performance optimization: Introduced ForkJoinPool for hub refresh. Default pool of 20 or processor count whichever is higher. Significant improvement in startup time.
- - Dropwizard version upgrade to 2.1.12

### CODE REVIEW

**Review by ToOnlyGaurav — COMMENTED** [review]
Please find my few observations.

File: ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java [hunk 1]
```diff
@@ -70,6 +73,8 @@ public void start() {
         try {
             RetryerBuilder.<Boolean>newBuilder()
                     .retryIfResult(r -> null == r || !r)
+                    .retryIfException()
+                    .withWaitStrategy(WaitStrategies.fixedWait(1, TimeUnit.SECONDS))
```
Comment (koushikr) [inline ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java:77]: Don't think the sleepTime should be hardcoded. Can we move this to an arg, in BaseServiceFinderBuilder where the updater is created, much like refreshTimeMs?

File: ranger-http/src/main/java/io/appform/ranger/http/serviceprovider/HttpNodeDataSink.java [hunk 2]
```diff
@@ -40,10 +40,12 @@
 public class HttpNodeDataSink<T, S extends HttpRequestDataSerializer<T>> extends HttpNodeDataStoreConnector<T> implements NodeDataSink<T, S> {
 
     private final Service service;
+    private final ObjectMapper mapper;
 
-    public HttpNodeDataSink(Service service, HttpClientConfig config, ObjectMapper mapper, OkHttpClient httpClient) {
-        super(config, mapper, httpClient);
+    public HttpNodeDataSink(Service service, HttpClientConfig config, ObjectMapper mapper, HttpCommunicator<T> httpClient) {
```
Comment (koushikr) [inline ranger-http/src/main/java/io/appform/ranger/http/serviceprovider/HttpNodeDataSink.java:45]: We can get away with the httpClientConfig also no, if there is a communicator interface already, we need that only to build the OkHttpClient and is not required otherwise, iirc. In that case we can get away with the clientConfig from multiple places.

Comment (koushikr) [inline ranger-http/src/main/java/io/appform/ranger/http/serviceprovider/HttpNodeDataSink.java:45]: We can get away with explicit httpClientConfig, when we have the communicator no? The communicator inherently should have an httpClientConfig.

Comment (koushikr) [reply ranger-http/src/main/java/io/appform/ranger/http/serviceprovider/HttpNodeDataSink.java:45]: Applicable in multiple places, perhaps?

File: ranger-http/src/main/java/io/appform/ranger/http/servicefinder/HttpCommunicator.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import io.appform.ranger.http.serde.HTTPResponseDataDeserializer;
+import okhttp3.OkHttpClient;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Interface for communicator to upstream
+ */
+public interface HttpCommunicator<T> extends AutoCloseable {
```
Comment (koushikr) [inline ranger-http/src/main/java/io/appform/ranger/http/servicefinder/HttpCommunicator.java:30]: To what I'm seeing, it is more of a NodeDataSource interface - functionally - than an HttpCommunicator. With both getServices and listNodes being part of that interface, further can be of types, zk, http, drove or whatever backend in the future. Practically combining both ServiceDataSource and NodeDataSource into a singular functional interface - which is understandable. And an httpCommunicator being a member of such an httpNodeDataSource is gettable. 

In this case, I don't see ServiceDataSource done away with. Is that the future idea? Since where you got the service list is where you will typically find your nodeList?

Comment (koushikr) [reply ranger-http/src/main/java/io/appform/ranger/http/servicefinder/HttpCommunicator.java:30]: The argument would only be on StaticServiceDataSource. That could be a property in the functional interface then?

Comment (koushikr) [reply ranger-http/src/main/java/io/appform/ranger/http/servicefinder/HttpCommunicator.java:30]: Either way, it is not counter-intuitive even if the interfaces stay as is, for now!

File: ranger-drove/src/main/java/io/appform/ranger/drove/common/DroveCommunicator.java [hunk 4]
```diff
@@ -30,6 +30,8 @@
 public interface DroveCommunicator extends AutoCloseable {
```
Comment (koushikr) [inline ranger-drove/src/main/java/io/appform/ranger/drove/common/DroveCommunicator.java:30]: Related to my comment on the HttpCommunicator - Increasingly this one now looks like a functional interface, with drove, http etc being implementations of it, and we can possibly do away with SDS in the future.

File: ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java (hunk 1 above)
Comment (koushikr) [reply ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java:77]: [NIT]

File: ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java [hunk 5]
```diff
@@ -125,21 +130,42 @@ private Void queryExecutor() {
...
+                if (null != nodeList) {
+                    log.debug("Updating nodeList of size: {} for [{}]", nodeList.size(),
+                              serviceRegistry.getService().getServiceName());
+                    val livenessCheckMaxAge = nodeDataSource.healthcheckZombieCheckThresholdTime(serviceRegistry.getService());
+                    //Remove all stale nodes before updating. This is done centrally to ensure some data sources
+                    //don't skip this check. Some control is still provided so that they can overload.
+                    serviceRegistry.updateNodes(FinderUtils.filterValidNodes(serviceRegistry.getService(), nodeList, livenessCheckMaxAge));
+                }
+                else {
+                    log.warn("Empty list returned from node data source. We are in a weird state. Keeping old list for {}",
```
[the remark this answers is not in the document]
Comment (koushikr) [reply ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java:148]: The nodeDataSource.refresh on callFailure - returns null, where the nodeList is retained as is, already. So that shouldn't be a problem. 

I don't know what you mean by nodeDataStore status? @ToOnlyGaurav 

TLDR: Old list is returned and maintained if the backend is not responding. And empty list is only returned by the nodeDataSource, when backend truly returns an empty list or the nodes are stale or unhealthy.

**Review by koushikr — COMMENTED** [review]
One more nit, 

There are a few args, which are hardcoded in code, like in HttpClientUtils, the connectionPool args, and on ServiceRegistryUpdater, sleepTime - We could move them to be provided during initialisation by the client instead. 

Doesn't apply to all, for example, ForkJoinPool args are fine.

File: ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java (hunk 1 above)
Comment (santanusinha) [reply ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java:77]: I just added  a wait time between retries. Otherwise it was doing retries with very little wait between them.

File: ranger-http/src/main/java/io/appform/ranger/http/serviceprovider/HttpNodeDataSink.java (hunk 2 above)
Comment (santanusinha) [reply ranger-http/src/main/java/io/appform/ranger/http/serviceprovider/HttpNodeDataSink.java:45]: this was already there and it goes to the base. Removed mapper from base and added it here.

File: ranger-drove/src/main/java/io/appform/ranger/drove/common/DroveCommunicator.java (hunk 4 above)
Comment (santanusinha) [reply ranger-drove/src/main/java/io/appform/ranger/drove/common/DroveCommunicator.java:30]: will think about refactoring later on.

File: ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java [hunk 6]
```diff
@@ -125,21 +130,42 @@ private Void queryExecutor() {
...
+            }
         }
-        else {
-            log.warn("Empty list returned from node data source. We are in a weird state. Keeping old list for {}",
+        if (!nodeDataSource.isActive() || callFailed) {
+            val currTime = System.currentTimeMillis();
+            log.warn("Node data source seems to be down. Keeping old list for {}." +
+                             " Will update timestamp to keep stale date relevant.",
                      serviceRegistry.getService().getServiceName());
+            serviceRegistry.updateNodes(serviceRegistry.nodeList()
```
[the remark this answers is not in the document]
Comment (koushikr) [reply ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java:164]: By extend scope, are you asking for another utility method that can update timestamps, in ServiceFinderUtils? @ToOnlyGaurav - that would be fine yeah, if it was what you are asking for.

[the remark this answers is not in the document]
Comment (ToOnlyGaurav) [reply ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java:164]: Yes, another method should be absolutely fine as it doing different job.

**Review by ToOnlyGaurav — APPROVED** [review]
LGTM from my side.
